### PR TITLE
Use pattern matching instead of checking lengths explicitly

### PR DESCRIPTION
**Pull request by LingMan**
This piece of code checks that there are exaclty two variants, one having
exactly one field, the other having exactly zero fields. If any of these
conditions is violated, it returns `None`. Otherwise it assigns that one
field's ty to `field_ty`.

Instead of fiddling with indices and length checks explicitly, use pattern
matching to simplify this.

@rustbot modify labels +C-cleanup +T-compiler

### CODE REVIEW

File: compiler/rustc_lint/src/types.rs [hunk 1]
```diff
@@ -783,25 +782,16 @@ crate fn repr_nullable_ptr<'tcx>(
...
-
-        let get_variant_fields = |index| &ty_def.variants[VariantIdx::new(index)].fields;
-        let variant_fields = [get_variant_fields(0), get_variant_fields(1)];
-        let fields = if variant_fields[0].is_empty() {
-            &variant_fields[1]
-        } else if variant_fields[1].is_empty() {
-            &variant_fields[0]
-        } else {
-            return None;
+        let field_ty = {
```
Comment (petrochenkov) [inline compiler/rustc_lint/src/types.rs:785]:
```suggestion
        let field_ty =
```
Nit: additional block around the `match` looks unnecessary.

**Comment by petrochenkov** [comment]
r=me with nit addressed.

File: compiler/rustc_lint/src/types.rs (hunk 1 above)
Comment (LingMan) [reply compiler/rustc_lint/src/types.rs]: Fixed.

**Pull request merged**